compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs

```diff
@@ -0,0 +1,170 @@
+///
+/// You can use `impl_tag` on structs and enums.
+/// You need to specify the type and all its possible values,
+/// which can only be paths with optional fields.
+///
+/// [`Tag`]: crate::tagged_ptr::Tag
+///
+/// # Examples
+///
+/// Basic usage:
+///
+/// ```
+/// use rustc_data_structures::{impl_tag, tagged_ptr::Tag};
+///
+/// #[derive(Copy, Clone, PartialEq, Debug)]
+/// enum SomeTag {
+///     A,
+///     B,
+///     X { v: bool },
+///     Y(bool, bool),
+/// }
+///
+/// impl_tag! {
+///     // The type for which the `Tag` will be implemented
+///     for SomeTag;
+///     // You need to specify the `{value_of_the_type} <=> {tag}` relationship
+///     SomeTag::A <=> 0,
+///     SomeTag::B <=> 1,
+///     // For variants with fields, you need to specify the fields:
+///     SomeTag::X { v: true  } <=> 2,
+///     SomeTag::X { v: false } <=> 3,
+///     // For tuple variants use named syntax:
+///     SomeTag::Y { 0: true,  1: true  } <=> 4,
+///     SomeTag::Y { 0: false, 1: true  } <=> 5,
+///     SomeTag::Y { 0: true,  1: false } <=> 6,
+///     SomeTag::Y { 0: false, 1: false } <=> 7,
+/// }
+///
+/// assert_eq!(SomeTag::A.into_usize(), 0);
+/// assert_eq!(SomeTag::X { v: false }.into_usize(), 3);
+/// assert_eq!(SomeTag::Y(false, true).into_usize(), 5);
+///
+/// assert_eq!(unsafe { SomeTag::from_usize(1) }, SomeTag::B);
+/// assert_eq!(unsafe { SomeTag::from_usize(2) }, SomeTag::X { v: true });
+/// assert_eq!(unsafe { SomeTag::from_usize(7) }, SomeTag::Y(false, false));
+/// ```
+///
+/// Structs are supported:
+///
+/// ```
+/// # use rustc_data_structures::impl_tag;
+/// #[derive(Copy, Clone)]
+/// struct Flags { a: bool, b: bool }
+///
+/// impl_tag! {
+///     for Flags;
+///     Flags { a: true,  b: true  } <=> 3,
+///     Flags { a: false, b: true  } <=> 2,
+///     Flags { a: true,  b: false } <=> 1,
+///     Flags { a: false, b: false } <=> 0,
+/// }
+/// ```
+///
+// This is supposed to produce a compile error, but does not,
+// see <https://github.com/rust-lang/rust/issues/110613> for more information.
+//
+// Using the same pattern twice results in a compile error:
+//
+// ```compile_fail
+// # use rustc_data_structures::impl_tag;
+// #[derive(Copy, Clone)]
+// struct Unit;
+//
+// impl_tag! {
+//     for Unit;
+//     Unit <=> 0,
+//     Unit <=> 1,
+// }
+// ```
+//
+// Using the same tag twice results in a compile error:
+//
+// ```compile_fail
+// # use rustc_data_structures::impl_tag;
+// #[derive(Copy, Clone)]
+// enum E { A, B };
+//
+// impl_tag! {
+//     for E;
+//     E::A <=> 0,
+//     E::B <=> 0,
+// }
+// ```
+//
+/// Not specifying all values results in a compile error:
+///
+/// ```compile_fail,E0004
+/// # use rustc_data_structures::impl_tag;
+/// #[derive(Copy, Clone)]
+/// enum E {
+///     A,
+///     B,
+/// }
+///
+/// impl_tag! {
+///     for E;
+///     E::A <=> 0,
+/// }
+/// ```
+#[macro_export]
+macro_rules! impl_tag {
+    (
+        for $Self:ty;
```

Comment:
```suggestion
        impl Tag for $Self:ty;
```
nit, makes it nicer to search for impls

Reply:
Done